Rewrite multi-layer stage 1 (C API) as net driver dispatcher (#792)
spatialos/UnrealGDK: SpatialGDK/Documentation/content/non-unreal-layers.md

>**Note:** Your external SpatialOS components must have an ID between 1000 and 2000 to be registered by the pipeline.

> You set up your game to receive network operations using the `SpatialNetDriver::AddOpCallback` function. This function is paramaterised with a SpatialOS component ID and a callback function reference with the following type:

Reviewer: ```suggestion
You set up your game to receive network operations using the `USpatialDispatcher::AddOpCallback` function. This function is paramaterised with a SpatialOS component ID and a callback function reference with the following type:
```

> You set up your game to receive network operations using the `USpatialDispatcher::AddOpCallback` function. This function is paramaterised with a SpatialOS component ID and a callback function reference with the following type:

Reviewer 2: ```suggestion
You set up your game to receive network operations using the `USpatialDispatcher::AddOpCallback` function. This function is paramaterized with a SpatialOS component ID and a callback function reference of the following types:
```
US-spelling.
I think the params are of two types?

> `const TFunction<void(Worker_ComponentId, const Worker_Op*)>`

Reviewer 2: ```suggestion
    ```
    const TFunction<void(Worker_ComponentId, const Worker_Op*)>
     ```
```

Trying to make it indent, as it makes it easier to scan and read the page with an ident here.

> where `Worker_ComponentId` and `Worker_Op` are types defined in the [Worker SDK in C’s API](https://docs.improbable.io/reference/latest/capi/reference).

Reviewer 2: ```suggestion
Here `Worker_ComponentId` and `Worker_Op` are types defined in the [Worker SDK in C’s API](https://docs.improbable.io/reference/latest/capi/reference).
```

> You'll need to register your callbacks before the Unreal worker connects to the SpatialOS runtime to avoid missing any operations.

Reviewer 3: I know this has been signed off for mvp, but this sort of thing worries me :(

Reviewer: Potential noob question :) Why does this worry you? because you expect users to want to add callbacks during the runtime of a game?

Reviewer 3: It's introducing quite a narrow contract that is easy to use incorrectly. It also implies that users will have to care about exact op semantics and interpret them correctly.

Reviewer 3: It's not wrong, it's just narrowing a contract because it was easy to make rather than because it makes something faster or helps a user in some broader context.

> You'll need to register your callbacks before the Unreal worker connects to the SpatialOS runtime to avoid missing any operations.

Reviewer 2: ```suggestion
You'll need to register your callbacks before the Unreal worker instance connects to the SpatialOS Runtime in order to avoid missing any operations.
```
Use instance to make it clear when we are talking about setting up a worker type or a worker instance doing something.
The SpatialOS Runtime is a Proper Noun.
"in order to" makes that sentence easier to read as there are two tos in there; one two too many to read with ease.

> You can deregister your callbacks at runtime using the `SpatialNetDriver::RemoveOpCallback` function and passing the `CallbackId` parameter returned by the corresponding call to `SpatialNetDriver::AddOpCallback`.

Reviewer: ```suggestion
You can deregister your callbacks at runtime using the `USpatialDispatcher::RemoveOpCallback` function and passing the `CallbackId` parameter returned by the corresponding call to `USpatialDispatcher::AddOpCallback`.
```

> You can deregister your callbacks at runtime using the `USpatialDispatcher::RemoveOpCallback` function and passing the `CallbackId` parameter returned by the corresponding call to `USpatialDispatcher::AddOpCallback`.

Reviewer 4: ```suggestion
You can deregister your callbacks using the `USpatialDispatcher::RemoveOpCallback` function and passing the `CallbackId` parameter returned by the corresponding call to `USpatialDispatcher::AddOpCallback`.
```

>**Note:** Your external SpatialOS components must have an ID between 1000 and 2000 to be registered by the pipeline.

You set up your game to receive network operations using the `USpatialDispatcher::AddOpCallback` functions. These overloaded functions are paramaterized with a `Worker_ComponentId` and a callback function reference that takes one of the following network operation types as an argument:

* `Worker_AddComponentOp`
* `Worker_RemoveComponentOp`
* `Worker_AuthorityChangeOp`
* `Worker_ComponentUpdateOp`
* `Worker_CommandRequestOp`
* `Worker_CommandResponseOp`

> `Worker_ComponentId` and each network operation type are defined in the [Worker SDK in C’s API](https://docs.improbable.io/reference/latest/capi/reference).

Author: nice catch

#### Receive data
You could receive and deserialize a component update and command request in your Unreal project code in the following way:
```
void OnAddComponent(Worker_ComponentId ComponentId, const Worker_Op* Op)
{
    Worker_ComponentUpdate Update = Op->component_update;
    
    // example deserialization logic
}

> 

Reviewer: Maybe we should have a small snippet for how to register with the spatial dispatcher here too, it may be obvious though. what do you think?

void OnCommandRequest(const Worker_CommandRequestOp& Op) override
{
    Worker_CommandRequest Request = Op->command_request;

    // example deserialization logic

    Cast<USpatialNetDriver>(World->GetNetDriver())->Connection->SendCommandResponse(Op.request_id, &Response);

> }

Reviewer: ```suggestion
```

#### Receive data
You could receive and deserialize a component update and command request in your Unreal project code in the following way:
```

> void ATPSGameMode::BeginPlay()

Reviewer 4: Slightly concerned about the race here between BeginPlay being called, and the NetDriver being setup. I believe if you test this code a switch off dedicated server it'll crash as there shouldn't be a net driver yet.

{
    USpatialDispatcher* Dispatcher = Cast<USpatialNetDriver>(GetWorld()->GetNetDriver())->Dispatcher;

>     Dispatcher->AddOpCallback(1337, [&](Worker_ComponentUpdateOp Op) {

Reviewer 3: If we change the other signatures remember to change this one too.

Reviewer 3: The worker SDK docs get away with '[&]' because it's in global scope. Here it's a bit iffy. Although I don't think this snippet actually needs to capture anything by ref. It looks like you might need to capture this, although it's probably better to get the world outside this scope.

        UWorld* World = GetWorld();

        uint32 PlayerCount = Schema_GetUint32(Schema_GetComponentUpdateFields(Op.update.schema_type), 1);
        UE_LOG(LogTPS, Verbose, TEXT("Received update, new player count: %d"), PlayerCount);

>         const FVector Location(0.0f, 0.0f, 0.0f);

Reviewer 4: ```suggestion
        const FVector Location = FVector::ZeroVector;
```

        const FRotator Rotation = FRotator::ZeroRotator;
        World->SpawnActor(CubeClass, &Location, &Rotation);
    });

    Dispatcher->AddOpCallback(1338, [&](Worker_CommandRequestOp Op) {
        Worker_CommandResponse Response = {};
        Response.component_id = 1338;
        Response.schema_type = Schema_CreateCommandResponse(1338, 1);
        Schema_Object* response_object = Schema_GetCommandResponseObject(Response.schema_type);

>         const char* text = "Here's my response.";

Reviewer 4: nit: It might be better if this example used an FString rather than a raw char*.

        Schema_AddBytes(response_object, 1, (const uint8_t*) TCHAR_TO_ANSI(*text), sizeof(char) * strlen(TCHAR_TO_ANSI(*text)));

        Cast<USpatialNetDriver>(GetWorld()->GetNetDriver())->Connection->SendCommandResponse(Op.request_id, &Response);
    });
}

> ```

Reviewer 2: Don't forget to update the metadata with the date and the update.

`2019-04-04 Page updated with full editorial review: updated methods for receiving network operations`
